File: solutions/047.permutations-ii/permutations-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const permuteUnique = nums => {
  nums.sort()
  let res = []

  const backtrack = (temp, visited) => {

    if (visited.every(e => e === 1)) {
      res.push(temp)
      return
    }

    for (let i = 0; i < nums.length; i++) {
      if (visited[i] === 1) continue


      let item = nums[i]
      if (nums[i] === nums[i - 1] && visited[i - 1]) continue

      visited[i] = 1
      backtrack([...temp, item], visited)
      visited[i] = undefined
    }

  }

  backtrack([], Array(nums.length).fill(0))

  return res
}

File: solutions/047.permutations-ii/permutations-ii.py
class Solution:
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        ret = []

        def traverse(tmp, visited):
            nonlocal nums, ret

            if sum(visited) == len(visited):
                ret.append(tmp)

            for i in range(len(visited)):

                if visited[i]:
                    continue

                if i > 0 and visited[i - 1] and nums[i] == nums[i - 1]:
                    continue

                visited[i] = 1
                traverse(tmp + [nums[i]], visited)
                visited[i] = 0

        traverse([], [0]*len(nums))

        return ret
